packages/connect-node/src/handler.ts

```diff
@@ -58,83 +52,148 @@ export type Handler = NodeHandler & {
+ * createHandlers() takes a service definition and a service implementation,
+ * and returns an array of Handler functions, one for each RPC.
  */
-export type ServiceImpl<T extends ServiceType> = {
-  [P in keyof T["methods"]]: MethodImpl<T["methods"][P]>;
-};
+export function createHandlers<T extends ServiceType>(
+  service: T,
+  implementation: ServiceImpl<T>,
+  options?: HandlerOptions
+): Handler[] {
+  return Object.entries(service.methods).map(([name, method]) => {
+    const i = implementation[name].bind(implementation);
+    return createHandler(service, method, i, options);
+  });
+}
 
-// prettier-ignore
 /**
- * MethodImpl is the signature of the implementation of an RPC.
+ * createHandler() takes an RPC definition and an RPC implementation, and
+ * returns a Handler function.
  */
-export type MethodImpl<M extends MI> =
-    M extends MI<infer I, infer O, MethodKind.Unary>           ? UnaryImpl<I, O>
-  : M extends MI<infer I, infer O, MethodKind.ServerStreaming> ? ServerStreamingImpl<I, O>
-  : M extends MI<infer I, infer O, MethodKind.ClientStreaming> ? ClientStreamingImpl<I, O>
-  : M extends MI<infer I, infer O, MethodKind.BiDiStreaming>   ? BiDiStreamingImpl<I, O>
-  : never;
+export function createHandler<M extends MethodInfo>(
+  service: ServiceType,
+  method: M,
+  impl: MethodImpl<M>,
+  options?: HandlerOptions
+): Handler {
+  const protocols = normalizeHandlerOptions(options);
+  const protocolHandlers = protocols.map((p) =>
+    Object.assign(p.createHandler(createImplSpec(service, method, impl)), p)
+  );
 
-interface MI<
-  I extends Message<I> = AnyMessage,
-  O extends Message<O> = AnyMessage,
-  K extends MethodKind = MethodKind
-> {
-  readonly kind: K;
-  readonly name: string;
-  readonly I: MessageType<I>;
-  readonly O: MessageType<O>;
-  readonly idempotency?: MethodIdempotency;
+  function handleAny(
+    req: http.IncomingMessage | http2.Http2ServerRequest,
+    res: http.ServerResponse | http2.Http2ServerResponse
+  ): void {
+    if (method.kind == MethodKind.BiDiStreaming && req.httpVersionMajor !== 2) {
+      return void endWithHttpStatus(res, 505, "Version Not Supported");
+    }
+    if (req.method !== "POST") {
+      // The gRPC-HTTP2, gRPC-Web, and Connect protocols are all POST-only.
+      return void endWithHttpStatus(res, 405, "Method Not Allowed");
+    }
+    const handleProtocol = protocolHandlers.find((p) =>
+      p.supportsMediaType(req.headers["content-type"] ?? "")
+    );
+    if (!handleProtocol) {
+      return void endWithHttpStatus(res, 415, "Unsupported Media Type");
+    }
+    handleProtocol(req, res).catch((reason) => {
+      // TODO need to handle rejections here, but it's unclear how exactly
+      // eslint-disable-next-line no-console
+      console.error("protocol handle failed", reason);
+    });
+  }
+  return Object.assign(handleAny, {
+    service,
+    method,
+    requestPath: createMethodUrl("https://example.com", service, method)
```

Comment:
Is this `example.com` url intended to be here?

Reply:
The function requires a valid `baseUrl` argument, even though we only want the path of the resulting URL in the end.
I agree it's irritating to read, and I hope we find a better way to do this.